drivers/sensor/bosch/bmm350/bmm350.c

```diff
@@ -0,0 +1,984 @@
+				 (((out_data[0] * ((data->mag_comp.cross_axis.cross_y_x *
+						    data->mag_comp.cross_axis.cross_z_y) -
+						   (data->mag_comp.cross_axis.cross_z_x *
+						    BMM350_MAG_COMP_COEFF_SCALING))) -
+				   (out_data[1] * ((data->mag_comp.cross_axis.cross_z_y *
+						    BMM350_MAG_COMP_COEFF_SCALING) -
+						   (data->mag_comp.cross_axis.cross_x_y *
+						    data->mag_comp.cross_axis.cross_z_x))))) /
+					 (((BMM350_MAG_COMP_COEFF_SCALING *
+					    BMM350_MAG_COMP_COEFF_SCALING) -
+					   data->mag_comp.cross_axis.cross_y_x *
+						   data->mag_comp.cross_axis.cross_x_y)));
+
+			out_data[0] = (int32_t)cr_ax_comp_x;
+			out_data[1] = (int32_t)cr_ax_comp_y;
+			out_data[2] = (int32_t)cr_ax_comp_z;
+		}
+		LOG_DBG("mag data %d %d %d\n", out_data[0], out_data[1], out_data[2]);
+
+		if (rslt == BMM350_OK) {
+			if ((data->axis_en & BMM350_EN_X_MSK) == BMM350_DISABLE) {
+				mag_temp_data->x = BMM350_DISABLE;
+			} else {
+				mag_temp_data->x = out_data[0];
+			}
+
+			if ((data->axis_en & BMM350_EN_Y_MSK) == BMM350_DISABLE) {
+				mag_temp_data->y = BMM350_DISABLE;
+			} else {
+				mag_temp_data->y = out_data[1];
+			}
+
+			if ((data->axis_en & BMM350_EN_Z_MSK) == BMM350_DISABLE) {
+				mag_temp_data->z = BMM350_DISABLE;
+			} else {
+				mag_temp_data->z = out_data[2];
+			}
+			mag_temp_data->temperature = out_data[3];
+		}
+	} else {
+		rslt = -EINVAL;
+	}
+	return rslt;
+}
+
+static int bmm350_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	struct bmm350_data *drv_data = dev->data;
+	struct bmm350_mag_temp_data mag_temp_data;
+
+	if (bmm350_get_compensated_mag_xyz_temp_data_fixed(dev, &mag_temp_data) < 0) {
+		LOG_ERR("failed to read sample");
+		return -EIO;
+	}
+	drv_data->mag_temp_data.x = mag_temp_data.x;
+	drv_data->mag_temp_data.y = mag_temp_data.y;
+	drv_data->mag_temp_data.z = mag_temp_data.z;
+	return 0;
+}
+
+/*
+ * ut change to Gauss
+ */
+static void bmm350_convert(struct sensor_value *val, int raw_val)
+{
+	val->val1 = raw_val / 100;
+	val->val2 = (raw_val % 100) * 10000;
+}
+
+static int bmm350_channel_get(const struct device *dev, enum sensor_channel chan,
+			      struct sensor_value *val)
+{
+	struct bmm350_data *drv_data = dev->data;
+
+	switch (chan) {
+	case SENSOR_CHAN_MAGN_X:
+		bmm350_convert(val, drv_data->mag_temp_data.x);
+		break;
+	case SENSOR_CHAN_MAGN_Y:
+		bmm350_convert(val, drv_data->mag_temp_data.y);
+		break;
+	case SENSOR_CHAN_MAGN_Z:
+		bmm350_convert(val, drv_data->mag_temp_data.z);
+		break;
+	case SENSOR_CHAN_MAGN_XYZ:
+		bmm350_convert(val, drv_data->mag_temp_data.x);
+		bmm350_convert(val + 1, drv_data->mag_temp_data.y);
+		bmm350_convert(val + 2, drv_data->mag_temp_data.z);
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+static uint8_t acc_odr_to_reg(const struct sensor_value *val)
+{
+	double odr = sensor_value_to_double((struct sensor_value *)val);
+
+	uint8_t reg = BMM350_DATA_RATE_100HZ;
+
+	if ((odr >= 0.78125) && (odr <= 1.5625)) {
```

Comment:
I really don't love these fuzzy ODR setters, but I know there's precedence in the tree for them.

What's weird about this idea is that you can set arbitrary ODRs and from the API response and call it may seem as if any ODR is possible because hey, it didn't fail!

The reality is the actual ODR is set to something else entirely... and unless you go read this attribute back you don't know.

Maybe ODR should be something we do more explicitly rather than with attributes. It's incredibly common, typically gets tied to some sample power mode (toggled/always on are common...) and really should give back the ODR in use rather than simply accepting any ODR and replying back with "Ok!"